design-proposals/golden-image-delivery-and-update-pipeline.md

```diff
@@ -0,0 +1,287 @@
+
+## Implementation Phases
+
+**Phase1** - CDI APIs
+* DataSource API/controller
+* DataImportCron API/controller
+
+**Phase2** - SSP + Common Templates
+* Common Templates transition to DataSources for DataVolumeTemplate source
+* SSP Operator creates DataImportCrons for automatic PVC import
+
+**Phase3** - HCO and UI
+* HCO Team integrates with SSP Operator’s API for setting volume import sources
+* UI Team transitions to detecting boot source using a linked DataSource rather than looking directly for a PVC of a specific name.
+
+# API Examples
+## VM + DataVolumeTemplate + DataSource as source
+
+```yaml
+apiVersion: cdi.kubevirt.io/v1beta1
+kind: DataVolume
+metadata:
+   name: my-fedora-vm-image
+spec:
+  pvc:
+    accessModes:
+    - ReadWriteOnce
+    resources:
+      requests:
+        storage: 2Gi
+  sourceRef:
+    kind: DataSource
+    name: fedora
+    namespace: golden-images 
+```
+
+## DataImportCron Example
+
+This example polls for image tags for the quay.io/kubevirt/fedora image and imports the latest image sorting tags based on semver. As new tags are introduced, the DataImportCron will import them as new PVCs and automatically manage updating the corresponding DataSource.
+
+* **source** specifies where to poll from
+* **managedDataSource** specifies the name of the corresponding DataSource this cron will manage. The controller should create this DataSource if it doesn't exist, and adopt ownership of the DataSource if it does already exist.
+* **schedule** specifies in cron format when and how often to look for new imports 
+* **garbageCollectOutdated** specifies whether old PVCs should be cleaned up after a new PVC is imported.
+
+On the status we have information pertaining to when the last imports occurred as well as conditions to indicate if the most recent import is successful or not.
+
+* **conditions** Indicate if the most recent import is successful or failing
+
+```yaml
+apiVersion: cdi.kubevirt.io/v1beta1
+kind: DataImportCron
+metadata:
+  name: fedora-image-import-cron
+  namespace: golden-images
+spec:
+  source:
+    registry:
+      url: docker://quay.io/kubevirt/fedora
+  schedule: "0 */12 * * *"
+  garbageCollectOutdated: true
+  managedDataSource: fedora
+status:
+  lastImportedPVC: fedora-v33
+  lastExecutionTimestamp: "01:01:01"
+  lastImport: "01:01:01"
+  conditions:
+  - status: "False"
+    message: "Unable to reach docker://quay.io/kubevirt/fedora"
+    type: ImportSuccessful
+```
+
+## DataSource Example
+
+This example shows how a DataSource uses a ready condition to indicate that the corresponding pvc exists and is populated. 
+
+The UI can use the ready condition to determine if a boot source exists for a DataSource or not. 
+
+```yaml
+apiVersion: cdi.kubevirt.io/v1beta1
+kind: DataSource
+metadata:
+  name: fedora
+spec:
+  source:
+    pvc:
+      namespace: golden-images 
+      name: fedora-32-001
+status:
+  conditions:
+  - status: "True"
+    type: Ready
```

Comment:
I don't think a DataSource can have an empty source, but it can point to a non-existent source.

For example, a Pod can be posted with a spec that requests attaching a PVC that doesn't exist. Once the PVC is available, the pod will get scheduled.

I see the same thing for the DataSource object, it could technically point to a PVC or DataVolume as a source, but there's no guarantee that source exists. In the event that the source does not exist, the `Ready` condition will be `false` with a message indicating it's waiting on the source to be accessible within the cluster.